File: packages/sdk/src/harness/loop/automation.py
"""
Automation - Simplified API for Trigger + Goal execution.

This module provides the Automation class, which is the primary
user-facing API for Loop Engineering Phase 2.
"""

from __future__ import annotations

import asyncio
import logging
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import TYPE_CHECKING, Any, Callable

from harness.loop.types import GoalConfig, GoalResult
from harness.triggers import (
    CronTrigger,
    IntervalTrigger,
    Trigger,
    TriggerAction,
    TriggerManager,
)

if TYPE_CHECKING:
    from harness.sdk.harness import AgentHarness

logger = logging.getLogger(__name__)

# Global manager singleton for Automation instances
_global_manager: TriggerManager | None = None


def get_global_manager(agent: "AgentHarness | None" = None) -> TriggerManager:
    """
    Get the global TriggerManager singleton.

    Args:
        agent: AgentHarness instance required for first initialization

    Returns:
        The global TriggerManager instance

    Raises:
        ValueError: If no agent is provided on first call
    """
    global _global_manager
    if _global_manager is None:
        if agent is None:
            raise ValueError(
                "Agent is required to initialize global TriggerManager. "
                "Call get_global_manager(agent=your_agent) first."
            )
        _global_manager = TriggerManager(agent)
    return _global_manager


def reset_global_manager() -> None:
    """
    Reset the global TriggerManager.

    This is primarily useful for testing.
    """
    global _global_manager
    _global_manager = None


class AutomationStatus(Enum):
    """
    Status of an Automation.

    Automations transition through these states during their lifecycle.
    """

    PENDING = "pending"       # Not started yet
    RUNNING = "running"       # Active and monitoring
    PAUSED = "paused"         # Temporarily paused
    STOPPED = "stopped"       # Permanently stopped
    ERROR = "error"           # Error state


@dataclass
class AutomationConfig:
    """
    Configuration for an Automation.

    An Automation combines a trigger with a goal, providing a simple
    way to set up scheduled or periodic goal execution.

    Attributes:
        name: Human-readable name for this automation
        goal: Goal description to execute when triggered

        # Trigger configuration (one of schedule, interval_seconds, or trigger)
        schedule: Cron expression (e.g., "0 9 * * *" for daily at 9:00)
        interval_seconds: Fixed interval in seconds
        trigger: Custom Trigger instance

        # Goal configuration
        workspace_dir: Working directory
        max_iterations: Maximum iterations per execution
        timeout_seconds: Execution timeout
        custom_verifier: Optional verification function

        # Output configuration
        skills: Skills to activate
        output_channels: Channels for results

        # Execution control
        max_retries: Retry attempts on failure
        retry_delay_seconds: Delay between retries
    """

    # Identity
    name: str

    # Goal
    goal: str

    # Trigger (one required)
    schedule: str | None = None
    interval_seconds: int | None = None
    trigger: Trigger | None = None

    # Goal configuration
    workspace_dir: str = "."
    max_iterations: int = 50
    timeout_seconds: int = 3600
    custom_verifier: Callable | None = None

    # Skills and output
    skills: list[str] = field(default_factory=list)
    output_channels: list[str] = field(default_factory=list)

    # Retry
    max_retries: int = 3
    retry_delay_seconds: float = 5.0

    def __post_init__(self):
        """Validate configuration."""
        if not self.name:
            raise ValueError("name is required")

        if not self.goal:
            raise ValueError("goal is required")

        # Ensure exactly one trigger is specified
        triggers = [
            self.schedule is not None,
            self.interval_seconds is not None,
            self.trigger is not None,
        ]
        if sum(triggers) == 0:
            raise ValueError(
                "One of schedule, interval_seconds, or trigger is required"
            )
        if sum(triggers) > 1:
            raise ValueError(
                "Only one of schedule, interval_seconds, or trigger can be specified"
            )


@dataclass
class AutomationResult:
    """
    Result of an Automation execution.

    Tracks the status and statistics of an automation.

    Attributes:
        automation_name: Name of the automation
        status: Current status
        goal_result: Result from last goal execution
        last_run: When the automation last ran
        run_count: Total number of executions
        error_count: Number of failed executions
        error_message: Last error message
    """

    automation_name: str
    status: AutomationStatus = AutomationStatus.PENDING
    goal_result: GoalResult | None = None
    last_run: datetime | None = None
    run_count: int = 0
    error_count: int = 0
    error_message: str | None = None


class Automation:
    """
    Automation - Simplified API for Trigger + Goal.

    Automation is the primary user-facing API for Loop Engineering Phase 2.
    It combines a trigger with a goal configuration, providing a simple
    interface for scheduled and periodic task execution.

    Example:
        ```python
        from harness.loop import Automation

        # Create a daily automation
        automation = Automation(
            name="daily-report",
            schedule="0 9 * * *",
            goal="Generate daily report and send to Slack",
            skills=["report-generation"],
        )

        # Start it
        await automation.start(agent)

        # Check status
        print(automation.status)

        # Stop it
        await automation.stop()
        ```

        ```python
        # Interval-based automation
        automation = Automation(
            name="health-check",
            interval_seconds=300,  # Every 5 minutes
            goal="Check system health",
        )
        await automation.start(agent)
        ```
    """

    def __init__(
        self,
        name: str,
        goal: str,
        schedule: str | None = None,
        interval_seconds: int | None = None,
        trigger: Trigger | None = None,
        **kwargs,
    ):
        """
        Create an Automation.

        Args:
            name: Human-readable name
            goal: Goal description to execute
            schedule: Cron expression for scheduled execution
            interval_seconds: Interval for periodic execution
            trigger: Custom trigger instance
            **kwargs: Additional AutomationConfig options
        """
        self.config = AutomationConfig(
            name=name,
            goal=goal,
            schedule=schedule,
            interval_seconds=interval_seconds,
            trigger=trigger,
            **kwargs,
        )

        self._trigger: Trigger | None = None
        self._agent: AgentHarness | None = None
        self._manager: TriggerManager | None = None
        self._trigger_id: str | None = None
        self._status = AutomationStatus.PENDING
        self._result = AutomationResult(automation_name=name)

    @classmethod
    def create(
        cls,
        name: str,
        goal: str,
        schedule: str | None = None,
        interval_seconds: int | None = None,
        **kwargs,
    ) -> "Automation":
        """
        Convenience factory method.

        Args:
            name: Automation name
            goal: Goal description
            schedule: Cron expression (optional)
            interval_seconds: Interval in seconds (optional)
            **kwargs: Additional options

        Returns:
            Automation instance
        """
        return cls(
            name=name,
            goal=goal,
            schedule=schedule,
            interval_seconds=interval_seconds,
            **kwargs,
        )

    @property
    def name(self) -> str:
        """Automation name."""
        return self.config.name

    @property
    def status(self) -> AutomationStatus:
        """Current status."""
        return self._status

    @property
    def result(self) -> AutomationResult:
        """Execution result (synced from TriggerManager if available)."""
        # Sync stats from TriggerManager if registered
        if self._manager and self._trigger_id:
            reg = self._manager.get_trigger(self._trigger_id)
            if reg:
                self._result.fire_count = reg.fire_count
                self._result.error_count = reg.error_count
                self._result.error_message = reg.last_error
                if reg.last_fired:
                    self._result.last_run = reg.last_fired
        return self._result

    @property
    def is_running(self) -> bool:
        """Check if automation is running."""
        return self._status == AutomationStatus.RUNNING

    async def start(
        self,
        agent: "AgentHarness",
        manager: TriggerManager | None = None,
    ) -> None:
        """
        Start the automation.

        Args:
            agent: AgentHarness instance to use for execution
            manager: Optional TriggerManager to register with.
                If not provided, uses the global manager.

        Raises:
            ValueError: If no valid trigger configuration
        """
        if self._status == AutomationStatus.RUNNING:
            logger.warning(f"Automation {self.name} is already running")
            return

        self._agent = agent

        # Use provided manager or get global manager
        if manager is None:
            manager = get_global_manager(agent)

        self._manager = manager

        # Create the trigger
        action = self._create_action()

        if self.config.schedule:
            self._trigger = CronTrigger(
                schedule=self.config.schedule,
                action=action,
                trigger_id=f"auto_{self.name}",
            )
        elif self.config.interval_seconds:
            self._trigger = IntervalTrigger(
                interval_seconds=self.config.interval_seconds,
                action=action,
                trigger_id=f"auto_{self.name}",
            )
        elif self.config.trigger:
            self._trigger = self.config.trigger
            self._trigger.action = action
        else:
            raise ValueError("No valid trigger configuration")

        # Register with manager instead of starting directly
        self._trigger_id = manager.register(self._trigger, action)

        # Ensure manager is running
        if not manager.is_running:
            await manager.start()

        self._status = AutomationStatus.RUNNING

        logger.info(
            f"Automation {self.name} started via TriggerManager "
            f"(trigger: {self._trigger.trigger_type.value})"
        )

    async def stop(self) -> None:
        """Stop the automation."""
        # Unregister from manager if registered
        if self._manager and self._trigger_id:
            self._manager.unregister(self._trigger_id)
            self._trigger_id = None
        elif self._trigger:
            # Fallback: direct trigger management (legacy behavior)
            await self._trigger.stop()
            self._trigger = None

        self._status = AutomationStatus.STOPPED
        logger.info(f"Automation {self.name} stopped after {self._result.run_count} runs")

    async def pause(self) -> None:
        """Pause the automation."""
        if self._manager and self._trigger_id:
            # Disable in manager (trigger stays registered but won't fire)
            self._manager.disable(self._trigger_id)
        elif self._trigger and self._trigger.is_running():
            # Fallback: direct trigger management
            await self._trigger.stop()

        self._status = AutomationStatus.PAUSED
        logger.info(f"Automation {self.name} paused")

    async def resume(self) -> None:
        """Resume a paused automation."""
        if self._status != AutomationStatus.PAUSED:
            logger.warning(f"Automation {self.name} is not paused")
            return

        if self._manager and self._trigger_id:
            # Re-enable in manager
            self._manager.enable(self._trigger_id)
            self._status = AutomationStatus.RUNNING
            logger.info(f"Automation {self.name} resumed")
        elif self._trigger and self._agent:
            # Fallback: direct trigger management
            await self._trigger.start(self._on_trigger)
            self._status = AutomationStatus.RUNNING
            logger.info(f"Automation {self.name} resumed")

    def _create_action(self) -> TriggerAction:
        """Create TriggerAction from configuration."""
        return TriggerAction(
            goal=self.config.goal,
            workspace_dir=self.config.workspace_dir,
            max_iterations=self.config.max_iterations,
            timeout_seconds=self.config.timeout_seconds,
            custom_verifier=self.config.custom_verifier,
            skills=self.config.skills,
            output_channels=self.config.output_channels,
            max_retries=self.config.max_retries,
            retry_delay_seconds=self.config.retry_delay_seconds,
        )

    def _on_trigger(self, event: Any) -> None:
        """
        Callback when trigger fires.

        Creates an async task to execute the goal.

        Args:
            event: TriggerEvent from the trigger
        """
        if self._status != AutomationStatus.RUNNING:
            return

        # Create async task for goal execution
        asyncio.create_task(self._execute_goal(event))

    async def _execute_goal(self, event: Any) -> None:
        """
        Execute the goal.

        Args:
            event: TriggerEvent that triggered execution
        """
        if not self._agent:
            logger.error(f"Automation {self.name} has no agent")
            return

        try:
            logger.info(f"Automation {self.name} executing goal")

            # Activate skills
            for skill_name in self.config.skills:
                try:
                    self._agent.activate_skill(skill_name)
                except Exception as e:
                    logger.warning(f"Failed to activate skill {skill_name}: {e}")

            # Build goal config
            goal_config = GoalConfig(
                description=self.config.goal,
                workspace_dir=self.config.workspace_dir,
                max_iterations=self.config.max_iterations,
                timeout_seconds=self.config.timeout_seconds,
                custom_verifier=self.config.custom_verifier,
            )

            # Execute
            result = await self._agent.run_goal(goal_config)

            # Update statistics
            self._result.goal_result = result
            self._result.last_run = datetime.now()
            self._result.run_count += 1

            if result.achieved:
                logger.info(
                    f"Automation {self.name} goal achieved in "
                    f"{result.total_iterations} iterations"
                )
            else:
                logger.warning(
                    f"Automation {self.name} goal not achieved: {result.status.value}"
                )

            # Handle output channels
            await self._handle_output(result)

        except Exception as e:
            self._result.error_count += 1
            self._result.error_message = str(e)
            self._status = AutomationStatus.ERROR
            logger.error(f"Automation {self.name} error: {e}")

    async def _handle_output(self, result: GoalResult) -> None:
        """
        Handle output to configured channels.

        Args:
            result: Goal execution result
        """
        for channel in self.config.output_channels:
            try:
                if channel == "console":
                    print(f"[{self.name}] {result.final_response}")
                elif channel == "log":
                    logger.info(f"[{self.name}] {result.final_response}")
                # Additional channels (slack, email) would be implemented here
            except Exception as e:
                logger.warning(f"Failed to output to {channel}: {e}")

    def __repr__(self) -> str:
        return (
            f"Automation(name={self.name!r}, status={self._status.value}, "
            f"runs={self._result.run_count})"
        )

File: packages/sdk-java/harness-sdk-loop/src/main/java/com/harness/loop/automation/Automation.java
package com.harness.loop.automation;

import com.harness.loop.GoalLoop;
import com.harness.loop.types.GoalConfig;
import com.harness.loop.types.GoalResult;
import com.harness.loop.types.GoalStatus;
import com.harness.loop.types.VerificationMethod;
import com.harness.types.LoopResult;
import com.harness.types.Session;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Consumer;

/**
 * Automation - Simplified API for scheduled goal execution.
 *
 * <p>Automation combines a trigger (cron or interval) with a goal configuration,
 * providing a simple interface for scheduled and periodic task execution.</p>
 *
 * <h2>Example</h2>
 * <pre>{@code
 * // Create a daily automation
 * Automation automation = Automation.builder()
 *     .name("daily-report")
 *     .goal("Generate daily report")
 *     .schedule("0 9 * * *")  // Daily at 9:00
 *     .build();
 *
 * // Start it
 * automation.start(agentRunner).join();
 *
 * // Check status
 * System.out.println(automation.getStatus());
 *
 * // Stop it
 * automation.stop().join();
 * }</pre>
 */
public class Automation {
    private static final Logger logger = LoggerFactory.getLogger(Automation.class);

    private final AutomationConfig config;
    private final AtomicReference<AutomationStatus> status;
    private final AutomationResult result;

    private GoalLoop.AgentRunner agent;
    private ScheduledExecutorService scheduler;
    private ScheduledFuture<?> scheduledFuture;
    private Consumer<Object> progressCallback;

    /**
     * Create an Automation with configuration.
     *
     * @param config Automation configuration
     */
    public Automation(AutomationConfig config) {
        this.config = config;
        this.status = new AtomicReference<>(AutomationStatus.PENDING);
        this.result = new AutomationResult(config.getName());
    }

    /**
     * Create an Automation with name and goal.
     *
     * @param name Automation name
     * @param goal Goal description
     */
    public Automation(String name, String goal) {
        this(AutomationConfig.builder()
                .name(name)
                .goal(goal)
                .intervalSeconds(60) // Default interval
                .build());
    }

    /**
     * Create a builder.
     */
    public static Builder builder() {
        return new Builder();
    }

    /**
     * Get the automation name.
     */
    public String getName() {
        return config.getName();
    }

    /**
     * Get the automation configuration.
     */
    public AutomationConfig getConfig() {
        return config;
    }

    /**
     * Get the current status.
     */
    public AutomationStatus getStatus() {
        return status.get();
    }

    /**
     * Get the execution result.
     */
    public AutomationResult getResult() {
        return result;
    }

    /**
     * Check if automation is running.
     */
    public boolean isRunning() {
        return status.get() == AutomationStatus.RUNNING;
    }

    /**
     * Set progress callback.
     *
     * @param callback Progress callback
     * @return this
     */
    public Automation withProgressCallback(Consumer<Object> callback) {
        this.progressCallback = callback;
        return this;
    }

    /**
     * Start the automation.
     *
     * @param agent Agent runner to execute goals
     * @return CompletableFuture that completes when started
     */
    public CompletableFuture<Void> start(GoalLoop.AgentRunner agent) {
        if (status.get() == AutomationStatus.RUNNING) {
            logger.warn("Automation {} is already running", getName());
            return CompletableFuture.completedFuture(null);
        }

        this.agent = agent;
        this.scheduler = Executors.newSingleThreadScheduledExecutor(r -> {
            Thread t = new Thread(r, "automation-" + getName());
            t.setDaemon(true);
            return t;
        });

        if (config.isCronBased()) {
            return startCronBased();
        } else {
            return startIntervalBased();
        }
    }

    private CompletableFuture<Void> startCronBased() {
        // Parse cron expression (simplified - just use interval for now)
        // Full cron support would require a library like Quartz
        String cron = config.getSchedule();
        long intervalMillis = parseCronToInterval(cron);

        logger.info("Starting automation {} with cron schedule: {} (interval: {}ms)",
                getName(), cron, intervalMillis);

        scheduledFuture = scheduler.scheduleAtFixedRate(
                this::executeGoal,
                0,
                intervalMillis,
                TimeUnit.MILLISECONDS
        );

        status.set(AutomationStatus.RUNNING);
        return CompletableFuture.completedFuture(null);
    }

    private CompletableFuture<Void> startIntervalBased() {
        long intervalMillis = config.getIntervalSeconds() * 1000L;

        logger.info("Starting automation {} with interval: {}s",
                getName(), config.getIntervalSeconds());

        scheduledFuture = scheduler.scheduleAtFixedRate(
                this::executeGoal,
                0,
                intervalMillis,
                TimeUnit.MILLISECONDS
        );

        status.set(AutomationStatus.RUNNING);
        return CompletableFuture.completedFuture(null);
    }

    /**
     * Stop the automation.
     *
     * @return CompletableFuture that completes when stopped
     */
    public CompletableFuture<Void> stop() {
        if (scheduledFuture != null) {
            scheduledFuture.cancel(false);
            scheduledFuture = null;
        }

        if (scheduler != null) {
            scheduler.shutdown();
            try {
                if (!scheduler.awaitTermination(5, TimeUnit.SECONDS)) {
                    scheduler.shutdownNow();
                }
            } catch (InterruptedException e) {
                scheduler.shutdownNow();
                Thread.currentThread().interrupt();
            }
            scheduler = null;
        }

        status.set(AutomationStatus.STOPPED);
        logger.info("Automation {} stopped after {} runs", getName(), result.getRunCount());

        return CompletableFuture.completedFuture(null);
    }

    /**
     * Pause the automation.
     */
    public void pause() {
        if (status.get() != AutomationStatus.RUNNING) {
            logger.warn("Automation {} is not running", getName());
            return;
        }

        if (scheduledFuture != null) {
            scheduledFuture.cancel(false);
        }

        status.set(AutomationStatus.PAUSED);
        logger.info("Automation {} paused", getName());
    }

    /**
     * Resume a paused automation.
     */
    public void resume() {
        if (status.get() != AutomationStatus.PAUSED) {
            logger.warn("Automation {} is not paused", getName());
            return;
        }

        // Restart scheduler
        this.scheduler = Executors.newSingleThreadScheduledExecutor(r -> {
            Thread t = new Thread(r, "automation-" + getName());
            t.setDaemon(true);
            return t;
        });

        if (config.isCronBased()) {
            long intervalMillis = parseCronToInterval(config.getSchedule());
            scheduledFuture = scheduler.scheduleAtFixedRate(
                    this::executeGoal,
                    0,
                    intervalMillis,
                    TimeUnit.MILLISECONDS
            );
        } else {
            scheduledFuture = scheduler.scheduleAtFixedRate(
                    this::executeGoal,
                    0,
                    config.getIntervalSeconds() * 1000L,
                    TimeUnit.MILLISECONDS
            );
        }

        status.set(AutomationStatus.RUNNING);
        logger.info("Automation {} resumed", getName());
    }

    /**
     * Execute the goal manually (outside of schedule).
     *
     * @return CompletableFuture with the goal result
     */
    public CompletableFuture<GoalResult> executeNow() {
        if (agent == null) {
            return CompletableFuture.failedFuture(
                    new IllegalStateException("Agent not set. Call start() first."));
        }

        return doExecuteGoal();
    }

    private void executeGoal() {
        if (status.get() != AutomationStatus.RUNNING) {
            return;
        }

        try {
            GoalResult goalResult = doExecuteGoal().join();

            if (goalResult.achieved()) {
                logger.info("Automation {} goal achieved in {} iterations",
                        getName(), goalResult.totalIterations());
            } else {
                logger.warn("Automation {} goal not achieved: {}",
                        getName(), goalResult.status().getValue());
            }

            handleOutput(goalResult);

        } catch (Exception e) {
            result.recordError(e.getMessage());
            result.setStatus(AutomationStatus.ERROR);
            logger.error("Automation {} error: {}", getName(), e.getMessage());
        }
    }

    private CompletableFuture<GoalResult> doExecuteGoal() {
        logger.info("Automation {} executing goal", getName());

        // Build goal config
        GoalConfig.Builder configBuilder = new GoalConfig.Builder()
                .description(config.getGoal())
                .workspaceDir(config.getWorkspaceDir())
                .maxIterations(config.getMaxIterations())
                .timeoutSeconds(config.getTimeoutSeconds());

        // Only use CUSTOM verification if customVerifier is provided
        if (config.getCustomVerifier() != null) {
            configBuilder.verificationMethod(VerificationMethod.CUSTOM);
            configBuilder.customVerifier(config.getCustomVerifier());
        }

        GoalConfig goalConfig = configBuilder.build();

        GoalLoop loop = new GoalLoop(agent, goalConfig, progressCallback);

        return loop.run().thenApply(goalResult -> {
            result.recordSuccess(goalResult);
            return goalResult;
        });
    }

    private void handleOutput(GoalResult goalResult) {
        for (String channel : config.getOutputChannels()) {
            try {
                switch (channel.toLowerCase()) {
                    case "console":
                        System.out.println("[" + getName() + "] " +
                                (goalResult.finalResponse() != null ? goalResult.finalResponse() : ""));
                        break;
                    case "log":
                        logger.info("[{}] {}", getName(), goalResult.finalResponse());
                        break;
                    default:
                        logger.warn("Unknown output channel: {}", channel);
                }
            } catch (Exception e) {
                logger.warn("Failed to output to {}: {}", channel, e.getMessage());
            }
        }
    }

    /**
     * Parse a simple cron expression to interval in milliseconds.
     *
     * <p>Supports simple formats:</p>
     * <ul>
     *   <li>"*\/N * * * *" - Every N minutes</li>
     *   <li>"0 * * * *" - Hourly</li>
     *   <li>"0 0 * * *" - Daily at midnight</li>
     * </ul>
     */
    private long parseCronToInterval(String cron) {
        String[] parts = cron.trim().split("\\s+");
        if (parts.length != 5) {
            logger.warn("Invalid cron format: {}, defaulting to 1 hour", cron);
            return TimeUnit.HOURS.toMillis(1);
        }

        String minute = parts[0];
        String hour = parts[1];

        // Every N minutes: "*/N * * * *"
        if (minute.startsWith("*/") && hour.equals("*")) {
            try {
                int interval = Integer.parseInt(minute.substring(2));
                return TimeUnit.MINUTES.toMillis(interval);
            } catch (NumberFormatException e) {
                // Fall through
            }
        }

        // Hourly: "0 * * * *"
        if (minute.equals("0") && hour.equals("*")) {
            return TimeUnit.HOURS.toMillis(1);
        }

        // Daily: "0 0 * * *" or specific hour
        if (minute.equals("0") && !hour.equals("*")) {
            return TimeUnit.HOURS.toMillis(24);
        }

        // Default: 1 hour
        logger.warn("Unsupported cron format: {}, defaulting to 1 hour", cron);
        return TimeUnit.HOURS.toMillis(1);
    }

    @Override
    public String toString() {
        return "Automation{" +
                "name='" + getName() + '\'' +
                ", status=" + status.get() +
                ", runs=" + result.getRunCount() +
                '}';
    }

    /**
     * Builder for Automation.
     */
    public static class Builder {
        private String name;
        private String goal;
        private String schedule;
        private Integer intervalSeconds;
        private String workspaceDir = ".";
        private int maxIterations = 50;
        private int timeoutSeconds = 3600;
        private java.util.function.Function<GoalResult, Boolean> customVerifier;
        private List<String> skills = new ArrayList<>();
        private List<String> outputChannels = new ArrayList<>();
        private int maxRetries = 3;
        private double retryDelaySeconds = 5.0;

        public Builder name(String name) {
            this.name = name;
            return this;
        }

        public Builder goal(String goal) {
            this.goal = goal;
            return this;
        }

        public Builder schedule(String schedule) {
            this.schedule = schedule;
            return this;
        }

        public Builder intervalSeconds(Integer intervalSeconds) {
            this.intervalSeconds = intervalSeconds;
            return this;
        }

        public Builder workspaceDir(String workspaceDir) {
            this.workspaceDir = workspaceDir;
            return this;
        }

        public Builder maxIterations(int maxIterations) {
            this.maxIterations = maxIterations;
            return this;
        }

        public Builder timeoutSeconds(int timeoutSeconds) {
            this.timeoutSeconds = timeoutSeconds;
            return this;
        }

        public Builder customVerifier(java.util.function.Function<GoalResult, Boolean> customVerifier) {
            this.customVerifier = customVerifier;
            return this;
        }

        public Builder skills(List<String> skills) {
            this.skills = new ArrayList<>(skills);
            return this;
        }

        public Builder addSkill(String skill) {
            this.skills.add(skill);
            return this;
        }

        public Builder outputChannels(List<String> outputChannels) {
            this.outputChannels = new ArrayList<>(outputChannels);
            return this;
        }

        public Builder addOutputChannel(String channel) {
            this.outputChannels.add(channel);
            return this;
        }

        public Builder maxRetries(int maxRetries) {
            this.maxRetries = maxRetries;
            return this;
        }

        public Builder retryDelaySeconds(double retryDelaySeconds) {
            this.retryDelaySeconds = retryDelaySeconds;
            return this;
        }

        public Automation build() {
            AutomationConfig.Builder configBuilder = AutomationConfig.builder()
                    .name(name)
                    .goal(goal)
                    .workspaceDir(workspaceDir)
                    .maxIterations(maxIterations)
                    .timeoutSeconds(timeoutSeconds)
                    .customVerifier(customVerifier)
                    .skills(skills)
                    .outputChannels(outputChannels)
                    .maxRetries(maxRetries)
                    .retryDelaySeconds(retryDelaySeconds);

            if (schedule != null) {
                configBuilder.schedule(schedule);
            } else if (intervalSeconds != null) {
                configBuilder.intervalSeconds(intervalSeconds);
            } else {
                throw new IllegalArgumentException("Either schedule or intervalSeconds is required");
            }

            return new Automation(configBuilder.build());
        }
    }
}
